src/coreclr/vm/typehandle.cpp

```diff
@@ -343,6 +344,99 @@ BOOL TypeHandle::IsCanonicalSubtype() const
     return (*this == TypeHandle(g_pCanonMethodTableClass)) || IsSharedByGenericInstantiations();
 }
 
+#ifndef DACCESS_COMPILE
+bool TypeHandle::IsManagedClassObjectPinned() const
+{
+    LIMITED_METHOD_DAC_CONTRACT;
+
+    // Function pointers are always mapped to typeof(IntPtr)
+    return !GetLoaderAllocator()->CanUnload() || IsFnPtrType();
+}
+
+void TypeHandle::AllocateManagedClassObject(RUNTIMETYPEHANDLE* pDest)
+{
+    REFLECTCLASSBASEREF refClass = NULL;
+
+    PTR_LoaderAllocator allocator = GetLoaderAllocator();
+
+    if (!allocator->CanUnload())
+    {
+        // Allocate RuntimeType on a frozen segment
+        // Take a lock here since we don't want to allocate redundant objects which won't be collected
+        CrstHolder exposedClassLock(AppDomain::GetMethodTableExposedClassObjectLock());
+
+        if (*pDest == NULL)
+        {
+            FrozenObjectHeapManager* foh = SystemDomain::GetFrozenObjectHeapManager();
+            Object* obj = foh->TryAllocateObject(g_pRuntimeTypeClass, g_pRuntimeTypeClass->GetBaseSize());
+            _ASSERTE(obj != NULL);
+            // Since objects are aligned we can use the lowest bit as a storage for "is pinned object" flag
+            _ASSERTE((((SSIZE_T)obj) & 1) == 0);
+            refClass = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(obj);
+            refClass->SetType(*this);
+            RUNTIMETYPEHANDLE handle = (RUNTIMETYPEHANDLE)obj;
+            // Set the bit to 1 (we'll have to reset it before use)
+            handle |= 1;
+            *pDest = handle;
+        }
+    }
+    else
+    {
+        GCPROTECT_BEGIN(refClass);
+        refClass = (REFLECTCLASSBASEREF)AllocateObject(g_pRuntimeTypeClass);
+        refClass->SetKeepAlive(allocator->GetExposedObject());
+        LOADERHANDLE exposedClassObjectHandle = allocator->AllocateHandle(refClass);
+        _ASSERTE((exposedClassObjectHandle & 1) == 0);
+        refClass->SetType(*this);
+
+        // Let all threads fight over who wins using InterlockedCompareExchange.
+        // Only the winner can set m_ExposedClassObject from NULL.
+        if (InterlockedCompareExchangeT(pDest, exposedClassObjectHandle, static_cast<LOADERHANDLE>(NULL)))
+        {
+            // GC will collect unused instance
+            allocator->FreeHandle(exposedClassObjectHandle);
+        }
+        GCPROTECT_END();
+    }
+}
+
+OBJECTREF TypeHandle::GetManagedClassObjectFromHandleFast(RUNTIMETYPEHANDLE handle)
+{
+    LIMITED_METHOD_CONTRACT;
+
+    // For a non-unloadable context, handle is expected to be either null (is not cached yet)
+    // or be a direct pointer to a frozen RuntimeType object
+
+    if (handle & 1)
+    {
+        // Clear the "is pinned object" bit from the managed reference
+        return (OBJECTREF)((handle >> 1) << 1); // C++ compiler is expected to emit "and reg, mask"
```

Comment:
Why is `(x >> 1) << 1` and hoping for an optimization to kick in better, than doing ` - 1` like what the existing code does?

Reply:
Ah, indeed! with `-1` it's now `lea     rax, [rdx-1]` instead of `and rdx, 0FFFFFFFFFFFFFFFEh`